### PR TITLE
Second attempt to bring `gfx-hal` to Amethyst

### PR DESCRIPTION
**Pull request by omni-viral**
This change is [Reviewable](https://reviewable.io/reviews/amethyst/amethyst/573)

### CODE REVIEW

File: amethyst_hal/src/factory.rs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    /// Create `Factory` from device and allocator.
+    pub(crate) fn new(
+        instance: B::Instance,
+        physical: B::PhysicalDevice,
+        device: B::Device,
+        allocator: SmartAllocator<B>,
+    ) -> Self {
+        #[allow(dead_code)]
+        fn is_send_sync<T: Send + Sync>() {}
+        is_send_sync::<Self>();
```
Comment (Xaeroxe) [inline amethyst_hal/src/factory.rs]: This might fit better in a test module.

File: amethyst_hal/src/factory.rs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+// #[derivative(Debug)]
+pub struct Factory<B: BackendEx> {
+    // #[derivative(Debug="ignore")]
+    instance: B::Instance,
+    physical: B::PhysicalDeviceEx,
+    // #[derivative(Debug="ignore")]
+    device: B::DeviceEx,
+    allocator: SmartAllocator<B>,
+    reclamation: ReclamationQueue<B>,
+    current: u64,
```
Comment (Xaeroxe) [inline amethyst_hal/src/factory.rs]: `current_frame` might be a better name.

File: amethyst_hal/src/system.rs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                queue.present(Some(&mut r.swapchain), Some(&job.release));
+
+                // Save job resources.
+                job.payload = Some(Payload {
+                    fence,
+                    acquire,
+                    pool,
+                });
+
+                // Enque frame.
```
Comment (Xaeroxe) [inline amethyst_hal/src/system.rs]: Correct spelling is enqueue

**Comment by torkleyy** [comment]
Looks really great! I'm glad we can let `xfg-rs` and `gfx-mem` do the hard work for us.

---

Reviewed 2 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[amethyst_hal/Cargo.toml, line 7 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dMOKlMy3IkYhyrM-F:-L5dMOKlMy3IkYhyrM-G:b2mbla5) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/Cargo.toml#L7)):*
> ```TOML
> 
> [features]
> gfx-vulkan = ["gfx-backend-vulkan"]
> ```

I don't think the feature names should have a `gfx-` prefix.

---

*[amethyst_hal/src/bundle.rs, line 19 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dMZcSfclZTrMz7hvf:-L5dMZcSfclZTrMz7hvg:b-fmxvpy) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/bundle.rs#L19)):*
> ```Rust
> pub struct RenderBundle<B> {
>     queues: usize,
>     pd: PhantomData<fn() -> B>,
> ```

No reason to use this workaround.

---

*[amethyst_hal/src/bundle.rs, line 59 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dMoFqcn-c4Ru4lVVF:-L5dMoFqcn-c4Ru4lVVG:b-ykmosb) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/bundle.rs#L59)):*
> ```Rust
>                 .filter(|family| family.queue_type() == QueueType::General)
>                 .next()
>                 .ok_or(format!("Can't find General queue family"))?;
> ```

Can we add an error kind as to not use strings as error everywhere?

---

*[amethyst_hal/src/bundle.rs, line 86 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dN2oFyLuH06f4vzQH:-L5dN2oFyLuH06f4vzQI:b-tkbunm) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/bundle.rs#L86)):*
> ```Rust
>         world.add_resource(factory);
> 
>         Ok(dispatcher.add_thread_local(system))
> ```

Can we make this a normal system for Dx12, Metal and Vulkan?

---

*[amethyst_hal/src/lib.rs, line 3 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dNeWmXcVKx0mz34tp:-L5dNeWmXcVKx0mz34tq:bskopyf) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/lib.rs#L3)):*
> ```Rust
> //!
> 
> #![deny(missing_docs)]
> ```

It's funny because here you make the compiler error in case of missing docs and above is the first doc line that is essentially missing. Please just make this a `warn`, otherwise experimenting with the implementation is hard because you need to add (useless) doc comments everywhere to make the project compile.

---

*[amethyst_hal/src/system.rs, line 1 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dO7wKIwGEUcYDf4hX:-L5dO7wKIwGEUcYDf4hY:bs0r341) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L1)):*
> ```Rust
> use std::collections::{HashMap, VecDeque};
> ```

I can't recommend using the std `HashMap`. For something small like `WindowId`, check out `fxhash`.

---

*[amethyst_hal/src/system.rs, line 104 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dOW5b2fkrhaKmMX9k:-L5dOW5b2fkrhaKmMX9l:bufltmn) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L104)):*
> ```Rust
>                         // Wait for job to finish.
>                         if !factory.wait_for_fence(&fence, !0) {
>                             panic!("Device lost or something");
> ```

Let's make this an `error!` call and a `return;`

---

*[amethyst_hal/src/system.rs, line 165 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dOwkhEgnc4ugnQuxO:-L5dOwkhEgnc4ugnQuxP:b-ya81ng) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L165)):*
> ```Rust
>                 group.queues[r.queue]
>                     .wait_idle()
>                     .expect("Device lost or something");
> ```

Same as above

---

*[amethyst_hal/src/system.rs, line 196 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dPoXIbg0hgTeJcjik:-L5dPoXIbg0hgTeJcjil:b-j9d7ye) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L196)):*
> ```Rust
>             .map(|f| f.started)
>             .min()
>             .unwrap();
> ```

If something goes wrong with the surfaces, I'm not sure this will work (as in not panic).

---

*[amethyst_hal/src/system.rs, line 314 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dQ1QnJj142P8Cq-e6:-L5dQ1Qo7QLX0XbZE_tl:bvs8d4a) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L314)):*
> ```Rust
>             depth: 0.0..1.0,
>         },
>         _ => panic!("Unsupported surface kind"),
> ```

Is this `unimplemented` or `unreachable`? Both ways, please change it to the respective macro.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/573#-:-L5dQEKz-TpVCsA-Nl_7:btizenm)*

**Comment by omni-viral** [comment]
Review status: 2 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[amethyst_hal/src/system.rs, line 314 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dQ1QnJj142P8Cq-e6:-L62nzdCWOewij68bOx6:b-ipz3wn) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L314)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

Is this `unimplemented` or `unreachable`? Both ways, please change it to the respective macro.
</blockquote></details>

It is bad input for the function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/573)*

**Comment by torkleyy** [comment]
Review status: 0 of 6 files reviewed at latest revision, 14 unresolved discussions.

---

*[amethyst_hal/src/factory.rs, line 177 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L6CDsQnugarASZS3Yz0:-L6CDsQnugarASZS3Yz1:b-yxhljv) ([raw file](https://github.com/amethyst/amethyst/blob/86b6d4583fa1f3cb1f6ba7296a31668a3e74cca7/amethyst_hal/src/factory.rs#L177)):*
> ```Rust
>     }
> 
>     pub(crate) fn get_surfaces<E>(&mut self, extend: &mut E)
> ```

Maybe something like `fill_surfaces` would be better?

---

*[amethyst_hal/src/system.rs, line 314 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dQ1QnJj142P8Cq-e6:-L6CGuhPoEWwqKU1--4U:b-ly8f51) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L314)):*
<details><summary><i>Previously, omni-viral (Zakarum) wrote…</i></summary><blockquote>

It is bad input for the function.
</blockquote></details>

I see. Is it even possible for `Surface` to return such a `Kind`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/573)*

File: amethyst_hal/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+extern crate derivative;
+extern crate gfx_hal as hal;
+extern crate gfx_memory as mem;
+#[macro_use]
+extern crate log;
+extern crate shred;
+extern crate specs;
+extern crate winit;
+extern crate xfg;
+
```
Comment (jojolepro) [inline amethyst_hal/src/lib.rs]: hal can't have opengl?

File: amethyst_hal/src/factory.rs [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+use {AmethystGraphBuilder, Buffer, Image};
+
+/// Extend backend trait with initialization method.
+pub trait BackendEx: Backend {
+    type Instance: Instance<Backend = Self> + Send + Sync;
+    fn init() -> Self::Instance;
+    fn create_surface(instance: &Self::Instance, window: &Window) -> Self::Surface;
+}
+
+#[cfg(feature = "gfx-vulkan")]
```
Comment (jojolepro) [inline amethyst_hal/src/factory.rs]: cargo build --example pong --features gfx-vulkan
```
[features]
gfx-vulkan = ["gfx-backend-vulkan"]
...
amethyst_hal = { path = "amethyst_hal", version = "0.1",feature="gfx-vulkan"}
[target.'cfg(not(target_os = "macos"))'.dependencies]
gfx-backend-vulkan = { version = "0.1", optional = true, git = "https://github.com/gfx-rs/gfx", rev = "68e80eaf"}
```
Didn't work.

File: Cargo.toml [hunk 6]
```diff
@@ -109,4 +110,10 @@ name = "appendix_a"
 path = "examples/appendix_a/main.rs"
 
 [workspace]
-members = ["amethyst_animation", "amethyst_gltf"]
```
[the remark this answers is not in the document]
Comment (Rhuagh) [reply Cargo.toml]: They depend on the old renderer

[the remark this answers is not in the document]
Comment (jojolepro) [reply Cargo.toml]: okay

**Comment by Rhuagh** [comment]
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[Cargo.toml, line 114 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L6ITiv7GUnnYyXeJMI-:-L6ITiv7GUnnYyXeJMI0:b-70o4rp) ([raw file](https://github.com/amethyst/amethyst/blob/86b6d4583fa1f3cb1f6ba7296a31668a3e74cca7/Cargo.toml#L114)):*
> ```TOML
> [workspace]
> 
> #[patch.'https://github.com/omni-viral/xfg-rs']
> ```

This needs to be fixed before we can merge.

---

*[amethyst_hal/Cargo.toml, line 7 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dMOKlMy3IkYhyrM-F:-L6ITpGz05dQZwLOsiwb:btd51a6) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/Cargo.toml#L7)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

I don't think the feature names should have a `gfx-` prefix.
</blockquote></details>

I agree.

---

*[amethyst_hal/src/factory.rs, line 168 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5JsB26sl4ig1Q_5NYW-r1-168:-L6IUTfGfr5Tn33JLCf0:b-7159op) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/factory.rs#L168)):*
<details><summary><i>Previously, Xaeroxe (Jacob Kiesel) wrote…</i></summary><blockquote>

This might fit better in a test module.
</blockquote></details>

Just put it in a `#[cfg(test)] mod tests {}` block at the end of the file?

---

*[amethyst_hal/src/factory.rs, line 141 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L6IUNZphFjo6l-Tzz7d:-L6IUNZphFjo6l-Tzz7e:b-4nttsk) ([raw file](https://github.com/amethyst/amethyst/blob/86b6d4583fa1f3cb1f6ba7296a31668a3e74cca7/amethyst_hal/src/factory.rs#L141)):*
> ```Rust
>     }
> 
>     /// Add new window to draw upon
> ```

Documentation wrong.

---

*[amethyst_hal/src/lib.rs, line 17 at r2](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5JsB26sl4ig1Q_5NYX-r2-17:-L6IUw4MoN8N-b6IaAvh:b-9frtfs) ([raw file](https://github.com/amethyst/amethyst/blob/86b6d4583fa1f3cb1f6ba7296a31668a3e74cca7/amethyst_hal/src/lib.rs#L17)):*
<details><summary><i>Previously, jojolepro wrote…</i></summary><blockquote>

hal can't have opengl?
</blockquote></details>

It can, but it's kind of incomplete, so we should add that back in once it gets a bit more mature.

---

*[amethyst_hal/src/system.rs, line 314 at r1](https://reviewable.io:443/reviews/amethyst/amethyst/573#-L5dQ1QnJj142P8Cq-e6:-L6IWxZrLYIZqYT5qd5j:bcspqcs) ([raw file](https://github.com/amethyst/amethyst/blob/5f0fef46bf6ba17e10e8d474e69e1a688d26f693/amethyst_hal/src/system.rs#L314)):*
<details><summary><i>Previously, torkleyy (Thomas Schaller) wrote…</i></summary><blockquote>

I see. Is it even possible for `Surface` to return such a `Kind`?
</blockquote></details>

`Kind` has other variants atleast, so it needs to be handled I guess?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/573)*

**Comment by omni-viral** [comment]
@jojolepro 
>  I converted the pong example to use hal

Could you post it somewhere?

**Comment by jojolepro** [comment]
Posted as gists, as I had trouble with git.
https://gist.github.com/jojolepro/9b2596d00feb802de171437c4ac33810

There's nothing really fancy. I just made things public and tried to integrate what I understood (which seems to be pretty limited :D ).

**Comment by torkleyy** [comment]
Reviewed 3 of 18 files at r3.
Review status: 4 of 19 files reviewed at latest revision, 20 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/amethyst/amethyst/573)*

**Comment by jondo2010** [comment]
Is this "attempt" dead?

**Comment by Rhuagh** [comment]
No, it's just being focused on other crates that will be used by the new renderer in the end.

**Comment by omni-viral** [comment]
This particular pr can be closed anyway.

**Pull request closed without merging**